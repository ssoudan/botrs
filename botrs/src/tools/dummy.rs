use std::fmt::Debug;

use llm_chain::tools::{Describe, Format, Tool, ToolDescription, ToolUseError};
use serde::{Deserialize, Serialize};

/// A tool that is called to test stuffs
#[derive(Default)]
pub struct DummyTool {}

/// A tool that is called to test stuffs
#[derive(Debug, Serialize, Deserialize)]
pub struct DummyToolInput {
    /// Well. MANDATORY.
    pub blah: String,
}

/// DummyToolOutput not very significant
#[derive(Serialize, Deserialize)]
pub struct DummyToolOutput {
    /// No much.
    pub something: String,
}

impl Describe for DummyToolInput {
    fn describe() -> Format {
        vec![("blah", "Well. MANDATORY.").into()].into()
    }
}

impl Describe for DummyToolOutput {
    fn describe() -> Format {
        vec![("something", "No much.").into()].into()
    }
}

impl DummyTool {
    fn invoke_typed(&self, input: &DummyToolInput) -> Result<DummyToolOutput, ToolUseError> {
        Ok(DummyToolOutput {
            something: input.blah.clone() + " and something else",
        })
    }
}

impl Tool for DummyTool {
    fn description(&self) -> ToolDescription {
        ToolDescription::new(
            "Dummy",
            "A tool to test stuffs.",
            "Use this to test stuffs.",
            DummyToolInput::describe(),
            DummyToolOutput::describe(),
        )
    }

    fn invoke(&self, input: serde_yaml::Value) -> Result<serde_yaml::Value, ToolUseError> {
        let input = serde_yaml::from_value(input)?;
        let output = self.invoke_typed(&input)?;
        Ok(serde_yaml::to_value(output)?)
    }
}
