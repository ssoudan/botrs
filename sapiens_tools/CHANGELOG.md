# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## v0.2.1 (2023-04-19)

### Documentation

 - <csr-id-2dc34812fa3afc6147fcd3f3b0bc5311b841ab9f/> Changelog

### Commit Statistics

<csr-read-only-do-not-edit/>

 - 9 commits contributed to the release.
 - 1 commit was understood as [conventional](https://www.conventionalcommits.org).
 - 0 issues like '(#ID)' were seen in commit messages

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **Uncategorized**
    - Changelog ([`2dc3481`](https://github.com/ssoudan/sapiens/commit/2dc34812fa3afc6147fcd3f3b0bc5311b841ab9f))
    - Adjusting changelogs prior to release of sapiens_derive v0.2.1, sapiens_tools v0.2.1, sapiens_cli v0.2.1 ([`3ea2039`](https://github.com/ssoudan/sapiens/commit/3ea2039b68192d4de6b9d370db54abcae054e3cc))
    - Release sapiens v0.2.1 ([`6d011b1`](https://github.com/ssoudan/sapiens/commit/6d011b16157847923433b870a6d57d5ad1b73438))
    - [+] SetStatusTool to control Hue lights. ([`e464518`](https://github.com/ssoudan/sapiens/commit/e4645184d43e99c7d90d7bc5ca91b43e3a034c8f))
    - [+] Derive(ProtoToolInvoke) ([`e383e20`](https://github.com/ssoudan/sapiens/commit/e383e2017123d2c2eee8a39ff34f50024001367a))
    - [+] rely on a published version of huelib-rs (named huelib2-rs). ([`cb91e27`](https://github.com/ssoudan/sapiens/commit/cb91e2795e5803f4c8ce8c41ed5605a006e83b15))
    - Release 0.2.0 ([`ab53b6c`](https://github.com/ssoudan/sapiens/commit/ab53b6c999892d82fbd9aed827a3a3bc1aee24a4))
    - [version] ([`43dd12d`](https://github.com/ssoudan/sapiens/commit/43dd12da54faaa1d580ff1e9c793b828592572b1))
    - [+] renaming ([`f664941`](https://github.com/ssoudan/sapiens/commit/f664941f2aba36cd9bce7493a19d030d2945bd50))
</details>

