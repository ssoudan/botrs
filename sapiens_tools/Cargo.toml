[package]
name = "sapiens_tools"
version= "0.2.2"
edition.workspace = true
authors.workspace = true
description = "The tools - Sapiens"
repository.workspace = true
license.workspace = true
publish.workspace = true
readme = "../README.md"
keywords.workspace = true
categories.workspace = true

[features]
default = ["wiki"]
# Hue lights
hue = ["dep:huelib2"]
# MediaWiki: Wikipedia, Wikidata
wiki = ["dep:mediawiki", "dep:serde_json"]

[dependencies]
sapiens = { path = "../sapiens", version = "^0.2.2" }
sapiens_derive = { path = "../sapiens_derive", version = "^0.2.2" }

huelib2 = { version = "0.13.3", optional = true }
mediawiki = { version = "0.2.10", optional = true  }

serde = { version = "1.0.160", features = ["derive"] }
serde_yaml = "0.9.21"
serde_json = { version = "1.0.96", optional = true }

tokio = { version = "1.27.0", features = ["macros"] }
async-trait = "0.1.68"

regex = "1.7.3"

pyo3 = { version = "0.18.3", features = [] }

tiktoken-rs = { version = "0.4.1", features = ["async-openai"] }
async-openai = "0.10.3"

convert_case = "0.6.0"

thiserror = "1.0.40"

[dev-dependencies]
indoc = "2"
serde_json = "1.0.96"
insta = { version = "1.29.0", features = ["yaml"] }
tokio = { version = "1.14.0", features = ["macros"] }
pyo3-asyncio = { version = "0.18.0", features = ["attributes", "tokio-runtime", "testing"]  }

[package.metadata.cargo-udeps.ignore]
normal = ["pyo3-asyncio"]

[[test]]
name = "e2e"
path = "tests/e2e.rs"
harness = false