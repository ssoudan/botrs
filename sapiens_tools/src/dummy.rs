use std::fmt::Debug;

use sapiens::tools::{Describe, ProtoToolDescribe, ProtoToolInvoke, ToolDescription, ToolUseError};
use sapiens_derive::{Describe, ProtoToolDescribe, ProtoToolInvoke};
use serde::{Deserialize, Serialize};

/// A tool that is called to test stuffs
#[derive(Debug, Default, ProtoToolDescribe, ProtoToolInvoke)]
#[tool(name = "Dummy", input = "DummyToolInput", output = "DummyToolOutput")]
#[allow(clippy::module_name_repetitions)]
pub struct DummyTool {}

/// A tool that is called to test stuffs
#[derive(Debug, Serialize, Deserialize, Describe)]
#[allow(clippy::module_name_repetitions)]
pub struct DummyToolInput {
    /// Well. MANDATORY.
    pub blah: String,
}

/// `DummyToolOutput` not very significant
#[derive(Serialize, Deserialize, Describe)]
#[allow(clippy::module_name_repetitions)]
pub struct DummyToolOutput {
    /// Not much.
    pub something: String,
}

impl DummyTool {
    #[tracing::instrument(skip(self))]
    async fn invoke_typed(&self, input: &DummyToolInput) -> Result<DummyToolOutput, ToolUseError> {
        Ok(DummyToolOutput {
            something: input.blah.clone() + " and something else",
        })
    }
}
