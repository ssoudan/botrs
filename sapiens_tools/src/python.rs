use std::rc::Rc;

use convert_case::{Case, Casing};
use pyo3::prelude::*;
use pyo3::types::{IntoPyDict, PyDict, PyFloat, PyList, PyTuple};
use sapiens::tools::{
    invoke_simple_from_toolbox, AdvancedTool, Describe, Format, FormatPart, ProtoToolDescribe,
    ProtoToolInvoke, ToolDescription, ToolUseError, Toolbox,
};
use sapiens_derive::{Describe, ProtoToolDescribe};
use serde::{Deserialize, Serialize};
use serde_yaml::Value;

const MAX_OUTPUT_SIZE: usize = 512;

/// A tool that runs sandboxed Python code. Use this to transform data.
/// - Only stdout and stderr are captured and made available (limited to 512B
/// total).
/// - To use other Tools from here: `input = {...}; output =
/// tools.tool_name(**input); print(output["field_xxx"])`. The `output` is an
/// object.
/// - List available tools with `tools.list()`. `tools` is already
/// imported.
/// - `open`|`exec` are forbidden.
/// - Limited libraries available: urllib3, requests, sympy, numpy,
/// BeautifulSoup4, feedparser.
/// - No PIP.
#[derive(Default, ProtoToolDescribe)]
#[tool(
    name = "SandboxedPython",
    input = "PythonToolInput",
    output = "PythonToolOutput"
)]
pub struct PythonTool {}

/// The input of the Python tool
#[derive(Serialize, Deserialize, Describe)]
pub struct PythonToolInput {
    /// The Python code to run. MANDATORY
    pub code: String,
}

/// The output of the Python tool
#[derive(Serialize, Deserialize, Describe)]
pub struct PythonToolOutput {
    /// The stdout output of the Python code.
    pub stdout: String,
    /// The stderr output of the Python code.
    pub stderr: String,
}

#[pyclass]
#[derive(Default)]
struct Logging {
    output: String,
}

#[pymethods]
impl Logging {
    fn write(&mut self, data: &str) {
        self.output.push_str(data);
    }
}

#[pyclass(unsendable)]
struct ToolsWrapper {
    toolbox: Rc<Toolbox>,
}

impl ToolsWrapper {
    fn new(toolbox: Rc<Toolbox>) -> Self {
        ToolsWrapper { toolbox }
    }
}

#[derive(thiserror::Error, Debug)]
enum PyConversionError {
    #[error("Invalid conversion: {error}")]
    InvalidConversion { error: String },
    #[error("dict key not serializable: {typename}")]
    DictKeyNotSerializable { typename: String },
    #[error("Invalid cast: {typename}")]
    InvalidCast { typename: String },
}

// inspired from https://github.com/mozilla-services/python-canonicaljson-rs/blob/62599b246055a1c8a78e5777acdfe0fd594be3d8/src/lib.rs#L87-L167
fn to_yaml(py: Python, obj: &PyObject) -> Result<Value, PyConversionError> {
    macro_rules! return_cast {
        ($t:ty, $f:expr) => {
            if let Ok(val) = obj.downcast::<$t>(py) {
                return $f(val);
            }
        };
    }

    macro_rules! return_to_value {
        ($t:ty) => {
            if let Ok(val) = obj.extract::<$t>(py) {
                return serde_yaml::to_value(val).map_err(|error| {
                    PyConversionError::InvalidConversion {
                        error: format!("{}", error),
                    }
                });
            }
        };
    }

    if obj.is_none(py) {
        return Ok(Value::Null);
    }

    return_to_value!(String);
    return_to_value!(bool);
    return_to_value!(u64);
    return_to_value!(i64);

    return_cast!(PyDict, |x: &PyDict| {
        let mut map = serde_yaml::Mapping::new();
        for (key_obj, value) in x.iter() {
            let key = if key_obj.is_none() {
                Ok("null".to_string())
            } else if let Ok(val) = key_obj.extract::<bool>() {
                Ok(if val {
                    "true".to_string()
                } else {
                    "false".to_string()
                })
            } else if let Ok(val) = key_obj.str() {
                Ok(val.to_string())
            } else {
                Err(PyConversionError::DictKeyNotSerializable {
                    typename: key_obj
                        .to_object(py)
                        .as_ref(py)
                        .get_type()
                        .name()
                        .map(|x| x.to_string())
                        .unwrap_or_else(|_| "unknown".to_string()),
                })
            };
            map.insert(Value::String(key?), to_yaml(py, &value.to_object(py))?);
        }
        Ok(Value::Mapping(map))
    });

    return_cast!(PyList, |x: &PyList| {
        let v = x
            .iter()
            .map(|x| to_yaml(py, &x.to_object(py)))
            .collect::<Result<Vec<_>, _>>()?;
        Ok(Value::Sequence(v))
    });

    return_cast!(PyTuple, |x: &PyTuple| {
        let v = x
            .iter()
            .map(|x| to_yaml(py, &x.to_object(py)))
            .collect::<Result<Vec<_>, _>>()?;
        Ok(Value::Sequence(v))
    });

    return_cast!(PyFloat, |x: &PyFloat| {
        Ok(Value::Number(serde_yaml::Number::from(x.value())))
    });

    // At this point we can't cast it, set up the error object
    Err(PyConversionError::InvalidCast {
        typename: obj
            .as_ref(py)
            .get_type()
            .name()
            .map(|x| x.to_string())
            .unwrap_or_else(|_| "unknown".to_string()),
    })
}

fn value_to_object(val: Value, py: Python<'_>) -> PyObject {
    match val {
        Value::Null => py.None(),
        Value::Bool(x) => x.to_object(py),
        Value::Number(x) => {
            let oi64 = x.as_i64().map(|i| i.to_object(py));
            let ou64 = x.as_u64().map(|i| i.to_object(py));
            let of64 = x.as_f64().map(|i| i.to_object(py));
            oi64.or(ou64).or(of64).expect("number too large")
        }
        Value::String(x) => x.to_object(py),
        Value::Sequence(x) => {
            let inner: Vec<_> = x.into_iter().map(|x| value_to_object(x, py)).collect();
            inner.to_object(py)
        }
        Value::Mapping(x) => {
            let iter = x
                .into_iter()
                .map(|(k, v)| (value_to_object(k, py), value_to_object(v, py)));
            IntoPyDict::into_py_dict(iter, py).into()
        }
        Value::Tagged(_) => panic!("tagged values are not supported"),
    }
}

/// Format of a field in a tool description
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SimpleFormat {
    /// Name of the field
    pub name: String,
    // pub r#type: String,
    /// Description of the field
    pub description: String,
}

impl From<FormatPart> for SimpleFormat {
    fn from(part: FormatPart) -> Self {
        SimpleFormat {
            name: part.key,
            // r#type: part.r#type,
            description: part.purpose,
        }
    }
}

impl ToPyObject for SimpleFormat {
    fn to_object(&self, py: Python<'_>) -> PyObject {
        let dict = PyDict::new(py);
        dict.set_item("name", self.name.to_object(py)).unwrap();
        // dict.set_item("type", self.r#type.to_object(py))
        //     .unwrap();
        dict.set_item("description", self.description.to_object(py))
            .unwrap();
        dict.into()
    }
}

/// A simplified version of the tool description
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SimpleToolDescription {
    /// Name of the tool
    pub name: String,
    /// Description of the tool
    pub description: String,
    /// Context utilization of the tool
    pub description_context: String,
    /// Input format of the tool
    pub input_format: Vec<SimpleFormat>,
    /// Output format of the tool
    pub output_format: Vec<SimpleFormat>,
}

impl From<ToolDescription> for SimpleToolDescription {
    fn from(desc: ToolDescription) -> Self {
        let input_format = desc
            .input_format
            .parts
            .into_iter()
            .map(|x| x.into())
            .collect();
        let output_format = desc
            .output_format
            .parts
            .into_iter()
            .map(|x| x.into())
            .collect();

        SimpleToolDescription {
            name: desc.name,
            description: desc.description,
            description_context: desc.description_context,
            input_format,
            output_format,
        }
    }
}

impl ToPyObject for SimpleToolDescription {
    fn to_object(&self, py: Python<'_>) -> PyObject {
        let dict = PyDict::new(py);
        dict.set_item("name", self.name.clone()).unwrap();
        dict.set_item("description", self.description.clone())
            .unwrap();
        dict.set_item("description_context", self.description_context.clone())
            .unwrap();
        dict.set_item("input_format", self.input_format.clone())
            .unwrap();
        dict.set_item("output_format", self.output_format.clone())
            .unwrap();
        dict.into()
    }
}

#[pymethods]
impl ToolsWrapper {
    // list all tools
    #[pyo3(signature = ())]
    fn list(&self, py: Python<'_>) -> PyResult<PyObject> {
        let tools = self.toolbox.describe();
        let tools = tools
            .into_values()
            .map(SimpleToolDescription::from)
            .collect::<Vec<_>>();
        let tools = tools.to_object(py);
        Ok(tools)
    }

    // invoke a tool
    #[pyo3(signature = (tool_name, input))]
    fn invoke(
        &self,
        py: Python<'_>,
        tool_name: &str,
        input: Option<&PyDict>,
    ) -> PyResult<PyObject> {
        // convert PyDict to a serde_yaml::Value
        let input = if let Some(input) = input {
            let input: PyObject = input.into();

            to_yaml(py, &input).map_err(|e| {
                pyo3::exceptions::PyException::new_err(format!("Invalid input: {}", e))
            })?
        } else {
            Value::default()
        };

        // println!("invoking tool {} with input {:?}", tool_name, input);

        let output =
            invoke_simple_from_toolbox(self.toolbox.clone(), tool_name, input).map_err(|e| {
                pyo3::exceptions::PyException::new_err(format!("Tool invocation failed: {}", e))
            })?;

        let output = value_to_object(output, py);

        Ok(output)
    }
}

impl PythonTool {
    fn invoke_typed(
        &self,
        toolbox: Option<Rc<Toolbox>>,
        input: &PythonToolInput,
    ) -> Result<PythonToolOutput, ToolUseError> {
        let mut code = input.code.clone();

        // check for forbidden keywords - with capture
        let re = regex::Regex::new(r"(exec|pip)").unwrap();
        if let Some(caps) = re.captures(&code) {
            return Err(ToolUseError::ToolInvocationFailed(format!(
                "Python code contains forbidden keywords such as {}",
                caps.get(0).unwrap().as_str()
            )));
        }

        let tools = toolbox.map(ToolsWrapper::new);

        // dynamically add functions to a `tools` module
        if let Some(tools) = &tools {
            let mut tool_class_code = String::new();

            tool_class_code.push_str("class Tools:\n");

            tool_class_code.push_str("    def __init__(self, toolbox):\n");
            tool_class_code.push_str("        self.toolbox = toolbox\n");

            for (name, description) in tools.toolbox.as_ref().describe() {
                let inputs_parts = description.input_format.parts;
                let inputs = inputs_parts
                    .iter()
                    .map(|f| f.key.clone())
                    .collect::<Vec<_>>()
                    .join(", ");
                let inputs = if inputs.is_empty() {
                    "".to_string()
                } else {
                    format!("(self, {})", inputs)
                };

                let dict = inputs_parts
                    .iter()
                    .map(|f| {
                        let name = &f.key;
                        format!("\"{}\": {}", name, name)
                    })
                    .collect::<Vec<_>>()
                    .join(", ");

                // in snake case
                tool_class_code.push_str(&format!(
                    "    def {}{}:\n        return self.toolbox.invoke(\"{}\", {{{}}})\n",
                    name.to_case(Case::Snake),
                    inputs,
                    name,
                    dict
                ));

                // in Pascal case
                tool_class_code.push_str(&format!(
                    "    def {}{}:\n        return self.toolbox.invoke(\"{}\", {{{}}})\n",
                    name.to_case(Case::Pascal),
                    inputs,
                    name,
                    dict
                ));

                // FUTURE(ssoudan) set input_format and output_format
            }

            // add list function
            tool_class_code.push_str("    def list(self):\n");
            tool_class_code.push_str("        return self.toolbox.list()\n");

            tool_class_code.push_str("tools = Tools(toolbox)\n");

            // prepend the tool class code to the user code
            code = format!("{}\n{}", tool_class_code, code);
        }

        // print!("{}", code);

        let res: PyResult<(String, String)> = Python::with_gil(|py| {
            // println!("Python version: {}", py.version());

            let globals = if let Some(tools) = tools {
                let tools_cell = PyCell::new(py, tools)?;
                [("toolbox", tools_cell)].into_py_dict(py)
            } else {
                PyDict::new(py)
            };

            // capture stdout and stderr
            let sys = py.import("sys")?;

            let stdout = Logging::default();
            let py_stdout_cell = PyCell::new(py, stdout)?;
            let py_stdout = py_stdout_cell.borrow_mut();
            sys.setattr("stdout", py_stdout.into_py(py))?;

            let stderr = Logging::default();
            let py_stderr_cell = PyCell::new(py, stderr)?;
            let py_stderr = py_stderr_cell.borrow_mut();
            sys.setattr("stderr", py_stderr.into_py(py))?;

            // FUTURE(ssoudan) pass something in

            // run code
            Python::run(py, &code, globals.into(), None)?;

            // NOFUTURE(ssoudan) get something out

            let stdout = py_stdout_cell.borrow().output.clone();
            let stderr = py_stderr_cell.borrow().output.clone();

            Ok((stdout, stderr))
        });

        let (stdout, stderr) = res.map_err(|e| {
            ToolUseError::ToolInvocationFailed(format!("Python code execution failed: {}", e))
        })?;

        Ok(PythonToolOutput { stdout, stderr })
    }
}

// impl ProtoToolDescribe for PythonTool {
//     fn description(&self) -> ToolDescription {
//         ToolDescription::new(
//             "SandboxedPython",
//             &format!("A tool that executes sandboxed Python code. Only stdout
// and stderr are captured and made available (limited to {}B total). ",
// MAX_OUTPUT_SIZE),             r#"Use this to transform data. To use other
// Tools from here: `input = {...}; output = tools.tool_name(**input);
// print(output["field_xxx"])`. List available tools with `tools.list()` -
// `tools` is already imported. The `output` is an object. open|exec are
// forbidden. Limited libraries available: urllib3, requests, sympy, numpy,
// BeautifulSoup4, feedparser. No PIP."#,
// PythonToolInput::describe(),             PythonToolOutput::describe(),
//         )
//     }
// }

impl ProtoToolInvoke for PythonTool {
    fn invoke(&self, input: serde_yaml::Value) -> Result<serde_yaml::Value, ToolUseError> {
        let input = serde_yaml::from_value(input)?;
        let output = self.invoke_typed(None, &input)?;

        // check the size of the output (stdout and stderr)
        let l = output.stdout.len() + output.stderr.len();
        if l > MAX_OUTPUT_SIZE {
            return Err(ToolUseError::ToolInvocationFailed(format!(
                "Python code produced too much output on stdout and stderr combined ({} bytes) - max is {}",
                l,
                MAX_OUTPUT_SIZE
            )));
        }

        Ok(serde_yaml::to_value(output)?)
    }
}

impl AdvancedTool for PythonTool {
    fn invoke_with_toolbox(
        &self,
        toolbox: Rc<Toolbox>,
        input: Value,
    ) -> Result<Value, ToolUseError> {
        let input = serde_yaml::from_value(input)?;
        let output = self.invoke_typed(Some(toolbox), &input)?;
        Ok(serde_yaml::to_value(output)?)
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;
    use pyo3::indoc::indoc;

    use super::*;
    use crate::dummy::DummyTool;

    #[test]
    fn test_python_tool() {
        let tool = PythonTool::default();
        let input = PythonToolInput {
            code: indoc! {
            r#"print('hello')
               t = toolbox.list()
               print("tools=", t)
               
               d = tools.dummy(blah="ahah")
               print("dummy=", d)
               
               "#}
            .to_string(),
        };
        let mut toolbox = Toolbox::default();
        toolbox.add_tool(DummyTool::default());
        let toolbox = Rc::new(toolbox);

        let output = tool.invoke_typed(Some(toolbox), &input).unwrap();
        assert_snapshot!(output.stdout);
        assert_snapshot!(output.stderr);
    }
}
