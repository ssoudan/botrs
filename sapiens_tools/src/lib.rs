//! Tools for sapiens

/// Hue tools
pub mod hue;

/// Tool to conclude a chain
pub mod conclude;

/// Tool to run some (limited) python
pub mod python;

#[cfg(test)]
/// Tool to test stuffs
pub(crate) mod dummy;

#[cfg(test)]
mod tests {
    use std::rc::Rc;

    use indoc::indoc;
    use sapiens::invoke_tool;
    use sapiens::tools::Toolbox;

    use crate::dummy::DummyTool;
    use crate::python::PythonTool;

    #[test]
    fn test_tool_invocation() {
        let data = indoc! {r#"
        # Action
        ```yaml        
        command: SandboxedPython
        input:
            code: |
                print("Hello world!")          
        ```
        "#};

        let mut toolbox = Toolbox::default();
        toolbox.add_advanced_tool(PythonTool::default());

        let toolbox = Rc::new(toolbox);

        let output = invoke_tool(toolbox, data).unwrap();
        assert_eq!(output, "stdout: |\n  Hello world!\nstderr: ''\n");
    }

    #[test]
    fn test_tool_invocation_in_python() {
        let data = indoc! {r#"
        # Action
        ```yaml        
        command: SandboxedPython
        input:
            code: |
                print("Hello world!")
                rooms = toolbox.invoke("Dummy", {"blah": "blah"})
                print(rooms)
                rooms = tools.dummy(blah="blah")
                print(rooms)          
        ```
        "#};

        let mut toolbox = Toolbox::default();
        toolbox.add_advanced_tool(PythonTool::default());
        toolbox.add_tool(DummyTool::default());

        let toolbox = Rc::new(toolbox);

        let output = invoke_tool(toolbox, data).unwrap();
        assert_eq!(
            output,
            "stdout: |\n  Hello world!\n  {'something': 'blah and something else'}\n  {'something': 'blah and something else'}\nstderr: ''\n"
        );
    }

    #[test]
    fn test_multiple_tool_invocations() {
        let data = indoc! {r#"
        # Action
        ```yaml        
        command: SandboxedPython
        input:
            code: |
                print("Hello world 1!")          
        ```
        
        # And another action
        ```yaml        
        command: SandboxedPython
        input:
            code: |
                print("Hello world 2!")          
        ```
        
        # And yet another action
        ```        
        command: SandboxedPython
        input:
            code: |
                print("Hello world 3!")          
        ```
        "#};

        let mut toolbox = Toolbox::default();
        toolbox.add_advanced_tool(PythonTool::default());

        let toolbox = Rc::new(toolbox);

        let output = invoke_tool(toolbox, data).unwrap();
        assert_eq!(output, "stdout: |\n  Hello world 1!\nstderr: ''\n");
    }
}
