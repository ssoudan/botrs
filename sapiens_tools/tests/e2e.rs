use std::rc::Rc;

use indoc::indoc;
use insta::assert_snapshot;
use sapiens::tools::{invoke_tool, Toolbox};
use sapiens_tools::dummy::DummyTool;
use sapiens_tools::python;
use sapiens_tools::python::PythonTool;

#[test]
fn test_tool_invocation() {
    let data = indoc! {r#"
        # Action
        ```yaml        
        command: SandboxedPython
        input:
            code: |
                print("Hello world!")          
        ```
        "#};

    let mut toolbox = Toolbox::default();
    toolbox.add_advanced_tool(PythonTool::default());

    let toolbox = Rc::new(toolbox);

    let (tool_name, res) = invoke_tool(toolbox, data);
    assert_eq!(tool_name, "SandboxedPython");
    let output = res.unwrap();
    assert_eq!(output, "stdout: |\n  Hello world!\nstderr: ''\n");
}

#[test]
fn test_tool_invocation_in_python() {
    let data = indoc! {r#"
        # Action
        ```yaml        
        command: SandboxedPython
        input:
            code: |
                print("Hello world!")
                rooms = toolbox.invoke("Dummy", {"blah": "blah"})
                print(rooms)
                rooms = tools.dummy(blah="blah")
                print(rooms)          
        ```
        "#};

    let mut toolbox = Toolbox::default();
    toolbox.add_advanced_tool(PythonTool::default());
    toolbox.add_tool(DummyTool::default());

    let toolbox = Rc::new(toolbox);

    let (tool_name, res) = invoke_tool(toolbox, data);
    assert_eq!(tool_name, "SandboxedPython");
    let output = res.unwrap();
    assert_eq!(
            output,
            "stdout: |\n  Hello world!\n  {'something': 'blah and something else'}\n  {'something': 'blah and something else'}\nstderr: ''\n"
        );
}

#[test]
fn test_multiple_tool_invocations() {
    let data = indoc! {r#"
        # Action
        ```yaml        
        command: SandboxedPython
        input:
            code: |
                print("Hello world 1!")          
        ```
        
        # And another action
        ```yaml        
        command: SandboxedPython
        input:
            code: |
                print("Hello world 2!")          
        ```
        
        # And yet another action
        ```        
        command: SandboxedPython
        input:
            code: |
                print("Hello world 3!")          
        ```
        "#};

    let mut toolbox = Toolbox::default();
    toolbox.add_advanced_tool(PythonTool::default());

    let toolbox = Rc::new(toolbox);

    let (tool_name, res) = invoke_tool(toolbox, data);
    assert_eq!(tool_name, "SandboxedPython");
    let output = res.unwrap();
    assert_snapshot!(output);
}

#[test]
fn test_python() {
    let mut toolbox = Toolbox::default();
    toolbox.add_tool(DummyTool::default());
    toolbox.add_advanced_tool(python::PythonTool::default());

    let toolbox = Rc::new(toolbox);

    let data = indoc! {
    r#"```yaml
       command: SandboxedPython
       input:
         code: |           
           args = {
               'blah': "hello"
           }
           output = tools.Dummy(**args)           
          
           something = output['something']                       

           print(f"And the result is: {something}")
       ```
    "#};

    let (tool_name, res) = invoke_tool(toolbox, data);
    assert_eq!(tool_name, "SandboxedPython");
    let output = res.unwrap();

    assert_snapshot!(output);
}

#[cfg(feature = "wiki")]
mod wiki {
    use indoc::indoc;

    #[test]
    fn test_wikidata_sparql() {
        let query = indoc! {
r#"
            PREFIX wd: <http://www.wikidata.org/entity/>
            PREFIX wdt: <http://www.wikidata.org/prop/direct/>
            PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
            
            SELECT ?country ?countryLabel ?capital ?capitalLabel
            WHERE {
              ?country wdt:P31 wd:Q6256 .         # ?country is an instance of a country (Q6256)
              ?country wdt:P36 ?capital .          # ?country has a capital (?capital)
              SERVICE wikibase:label {
                bd:serviceParam wikibase:language "en" .    # Use English labels
                ?country rdfs:label ?countryLabel .
                ?capital rdfs:label ?capitalLabel .
              }
            }
            ORDER BY ?countryLabel
            LIMIT 10
        "#
        };

        let api = mediawiki::api_sync::ApiSync::new("https://www.wikidata.org/w/api.php").unwrap(); // Will determine the SPARQL API URL via site info data
        let res = api.sparql_query(query).unwrap();
        println!("{}", serde_json::to_string_pretty(&res).unwrap());
    }

    #[test]
    fn test_wikipedia() {
        let api = mediawiki::api_sync::ApiSync::new("https://en.wikipedia.org/w/api.php").unwrap();

        // Query parameters
        let params = api.params_into(&[
            ("action", "query"),
            ("prop", "extracts|explaintext|exintro"),
            ("titles", "Albert Einstein"),
        ]);

        let res = api.get_query_api_json_all(&params).unwrap();

        // Print the result
        println!("{}", serde_json::to_string_pretty(&res).unwrap());
    }
}
