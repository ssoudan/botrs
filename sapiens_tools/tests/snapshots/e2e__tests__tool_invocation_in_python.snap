---
source: sapiens_tools/tests/e2e.rs
expression: output
---
stdout: |
  Hello world!
  {'something': 'blah and something else'}
  {'something': 'blah and something else'}
stderr: ''

